src/main/java/by/custom/utilcalculator/domain/UserProgress.java

```diff
@@ -5,155 +5,200 @@
+    private Weight weight = null;
+    private TrailerO4Type trailersO4Type = null;
     private final String chatID;
-    private Step currentQuestion;
+    private final List<Command> userPath;
 
     public UserProgress(final String chatID) {
         this.chatID = chatID;
+        userPath = new ArrayList<>();
+    }
+
+    private void addUserStatusToPath(final StepsIndicator stepsIndicator) {
+        final Map<StepsIndicator, Command> fieldsToCommands = CommandTree.init().getFieldsToCommands();
+        Command command = fieldsToCommands.get(stepsIndicator);
+        Class<? extends StepsIndicator> stepsIndicatorClass = stepsIndicator.getClass();
+        boolean needAddCommand = true;
+
+        for (int i = 0; i < userPath.size(); i++) {
+            Command localCommand = userPath.get(i);
+            for (Map.Entry<StepsIndicator, Command> entry : fieldsToCommands.entrySet()) {
+                if (entry.getValue().equals(localCommand)) {
+                    if (entry.getKey().getClass().equals(stepsIndicatorClass)) {
+                        needAddCommand = false;
+                        userPath.set(i, command);
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (needAddCommand) {
+            userPath.add(command);
+        }
     }
 
-    public TransportType getTransportType(){
-        return transportType;
+    public Step getNextStep() {
+        return TreeMediator.getNextStep(this);
+    }
+
+    public GeneralTransportType getGeneralTransportType() {
+        return generalTransportType;
     }
 
     public CountryOrigin getCountryOrigin() {
         return countryOrigin;
     }
 
-    public Command[] getUserPath() {
-        final Command[] userPath = new Command[6];
-        final Map<StepsIndicator, Command> fieldsToCommands = CommandTree.getInstance().getFieldsToCommands();
-        userPath[0] = fieldsToCommands.get(getTransportType()); //0 cell contains type of transport
-        userPath[1] = fieldsToCommands.get(getCountryOrigin()); //1 cell contains country origin
-        userPath[2] = fieldsToCommands.get(getOwnersType()); //2 cell contains owners type
-        userPath[3] = fieldsToCommands.get(getCarAge()); //3 cell contains car age
-        userPath[4] = fieldsToCommands.get(getTypeOfEngine()); //4 cell contains type of engine
-        userPath[5] = fieldsToCommands.get(getVolumeOfEngine()); //5 cell contains engine volume
+    public List<Command> getUserPath() {
         return userPath;
     }
 
-    public void setTransportType(final TransportType transportType) {
-        final int stepID = 0;
-        cleanStepsAfterCurrent(stepID);
-        currentQuestion = Step.TRANSPORT_TYPE;
-        this.transportType = transportType;
+    public TrailerO4Type getTrailerO4Type() {
+        return trailersO4Type;
+    }
+
+    public void setTrailerO4Type(final TrailerO4Type trailersO4Type) {
+        cleanStepsAfterCurrentM1Branch(3);
```

Comment:
You have to calculate the step number. I expect to see something like this.
```suggestion
        var newStep = TreeMediator.getNextStep(this);
        cleanStepsAfterCurrent(newStep.number);
        this.trailersO4Type = trailersO4Type;
        updateUserPath(trailersO4Type);
```

Reply:
In my head we don't need cleanSteps here future - it should works automatically. I'm going to write to you my solution in telegram and I'll be happy to discuss it with you.